File: sty_algorithm/src/main/java/com/sty/algorithm/leetcode/editor/cn/[541]反转字符串 II.java
package com.sty.algorithm.leetcode.editor.cn;

//leetcode submit region begin(Prohibit modification and deletion)
class Solution541 {
    public String reverseStr(String s, int k) {
        char[] charArray = s.toCharArray();

        for (int i = 0; i < s.length(); i += 2 * k) {
            reverse(charArray, i, Math.min(i + k, s.length()) - 1);
        }

        return new String(charArray);
    }

    public void reverse(char[] charArray, int left, int right) {
        while (left <= right) {
            char temp = charArray[right];
            charArray[right] = charArray[left];
            charArray[left] = temp;
            left++;
            right--;
        }
    }


}
//leetcode submit region end(Prohibit modification and deletion)

File: sty_algorithm/src/main/scala/com/sty/algorithm/leetcode/editor/cn/[541]反转字符串 II.scala

//leetcode submit region begin(Prohibit modification and deletion)
object Solution {

  def reverseStr(s: String, k: Int): String = {
    val array = s.toCharArray
    for (i <- s.indices by 2 * k) {
      // 如果i+k大于了res的长度，则需要全部翻转
      if (i + k > s.length) {
        reverseStr(array, i, s.length - 1)
      } else {
        reverseStr(array, i, i + k - 1)
      }
    }

    new String(array)
  }

  private def reverseStr(arr: Array[Char], left: Int, right: Int): Array[Char] = {
    var (l, r) = (left, right)
    while (l < r) {
      val temp = arr(r)
      arr(r) = arr(l)
      arr(l) = temp
      l += 1
      r -= 1
    }

    arr
  }
}
//leetcode submit region end(Prohibit modification and deletion)
